### PR TITLE
p2p: upgrade libp2p

### PR DESCRIPTION
**Pull request by algorandskiy**
## Summary

Upgraded go-libp2p, go-libp2p-pubsub, go-libp2p-pubsub to the latest versions supporting go1.21 (go-libp2p is one patch version behind).

Fixed `*basichost.BasicHost` casting since `libp2p.New` does not return `*basichost.BasicHost` anymore.
Fixed private addresses test:
  * `/ip6/2001:db8::/ipcidr/32` is supported natively by multiaddr lib (i.e. fixed a known TODO item)
  * `/ip6/200:11::/tcp/1234` ended up to be a bad example for testing public IPv6 addresses after some improvements in multiaddr

## Extra upgrades

Upgrading libp2p also triggered some other dependencies upgrades, notable `golang.org/x/crypto` v0.21 -> v0.26.
go-algorand directly uses  `acme/autocert`,, `blake2b`, `sha3`, `hkdf`, `nacl/secretbox` and `scrypt` submodules.
The 4 latter (`sha3` ... `scrypt`) might potentially affect on-chain data through AVM and kmd.

[Version history v0.21 -> v0.26](https://github.com/golang/crypto/compare/v0.21.0...v0.26.0) shows only `sha3` was changed and both changes (xor, buf) looks good/safe.

## Test Plan

Existing unit tests. Performance test showed some regression but we have to upgrade anyway.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -147,44 +148,62 @@ require (
...
+	github.com/pion/randutil v0.1.0 // indirect
+	github.com/pion/rtcp v1.2.14 // indirect
+	github.com/pion/rtp v1.8.9 // indirect
+	github.com/pion/sctp v1.8.33 // indirect
+	github.com/pion/sdp/v3 v3.0.9 // indirect
+	github.com/pion/srtp/v2 v2.0.20 // indirect
+	github.com/pion/stun v0.6.1 // indirect
+	github.com/pion/transport/v2 v2.2.10 // indirect
+	github.com/pion/turn/v2 v2.1.6 // indirect
+	github.com/pion/webrtc/v3 v3.3.0 // indirect
```
Comment (gmalouf) [inline go.mod:172]: I think these are strictly used by libp2p, still a lot of new dependencies.

File: network/p2p/p2p.go [hunk 2]
```diff
@@ -281,9 +280,13 @@ func (s *serviceImpl) dialNode(ctx context.Context, peer *peer.AddrInfo) error {
 
 // AddrInfo returns the peer.AddrInfo for self
 func (s *serviceImpl) AddrInfo() peer.AddrInfo {
+	addrs, err := s.host.Network().InterfaceListenAddresses()
+	if err != nil {
+		s.log.Errorf("failed to get listen addresses: %v", err)
+	}
 	return peer.AddrInfo{
 		ID:    s.host.ID(),
-		Addrs: s.host.(*basichost.BasicHost).AllAddrs(), // fetch all addresses, including private ones
```
Comment (gmalouf) [inline network/p2p/p2p.go:286]: Are we sure the new call is returning the same addresses/format as this legacy one was?

**Review by gmalouf — APPROVED** [review]
I'm good with this in general, would like @jannotti to skim on thoughts around any libraries affecting AVM.

**Review by jannotti — APPROVED** [review]
Seems like the crypto changes are minimal, so I don't see any problems.

File: go.mod (hunk 1 above)
Comment (algorandskiy) [reply go.mod:172]: yes, `pion` stuff is used in `go-libp2p`

File: network/p2p/p2p.go (hunk 2 above)
Comment (algorandskiy) [reply network/p2p/p2p.go:286]: `AllAddrs` is listen addrs + `HasDiscoveredNAT` or `OwnObservedAddrs`

`InterfaceListenAddresses` is only listen addrs.

Our usages are:
```
// p2pNetwork
localAddr, has := n.Address()

// node
func (node *AlgorandFullNode) Start() {
	//...
	node.config.NetAddress, _ = node.net.Address()
	//...
}

func (node *AlgorandFullNode) ListeningAddress() (string, bool) {
	return node.net.Address()
}
```
So we do not expect anything other than a listen address actually, so it fits well.

**Pull request merged**